internal/app/upstream/client_test.go

```diff
@@ -358,6 +362,60 @@ func TestOpenStreamShouldRetryIfSendFails(t *testing.T) {
 	})
 }
 
+func TestStreamShouldRetryWhenTimeoutMet(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	responseChan := make(chan *v2.DiscoveryResponse)
+	scope := stats.NewMockScope("mock")
+	client := createMockClientWithResponse(ctx, time.Second, responseChan, func(m interface{}) error {
+		return nil
+	}, scope, 1)
+
+	_, done := client.OpenStream(
+		transport.NewRequestV2(&v2.DiscoveryRequest{
+			TypeUrl: resource.ListenerType,
+			Node:    &core.Node{},
+		}), "aggregated_key")
+	defer done()
+	for {
+		var val int64 = 0
+		if scope.Snapshot().Counters() != nil &&
+			scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"] != nil {
+			val = scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"].Value()
+		}
+		if val > 0 {
+			break
+		}
+	}
```

Comment:
If the test condition fails, this loop will cause the test to hang indefinitely is that correct? Can we make this test automatically fail if a set duration has passed?

Reply:
Yes this is correct. The latest commit should check for this condition and fail the test after 5s if the success condition isn't met.